File: python/paceometer_math/speed_limit_parsing.py
"""Pure OSM maxspeed-tag parsing -- a Python mirror of
js/math/speedLimitParsing.js. Dev-only reference/test oracle; see
pace_math.py's module docstring.
"""
import re

from .geo_math import haversine_meters

_KMH_RE = re.compile(r"^(\d+(?:\.\d+)?)\s*km/?h$")
_MPH_RE = re.compile(r"^(\d+(?:\.\d+)?)\s*mph$")
_BARE_RE = re.compile(r"^(\d+(?:\.\d+)?)$")


def parse_maxspeed_tag(raw):
    if not isinstance(raw, str):
        return None
    trimmed = raw.strip().lower()

    kmh_match = _KMH_RE.match(trimmed)
    if kmh_match:
        return float(kmh_match.group(1)) * 0.621371

    mph_match = _MPH_RE.match(trimmed)
    if mph_match:
        return float(mph_match.group(1))

    bare_match = _BARE_RE.match(trimmed)
    if bare_match:
        return float(bare_match.group(1))

    return None


def extract_maxspeed_mph(data):
    elements = (data or {}).get("elements") or []
    for element in elements:
        tags = element.get("tags") or {}
        mph = parse_maxspeed_tag(tags.get("maxspeed"))
        if mph is not None:
            return mph
    return None


def cached_speed_limit_near(cache, coords, timestamp, max_age_ms, radius_meters):
    for entry in reversed(cache):
        if timestamp - entry["timestamp"] > max_age_ms:
            continue
        if haversine_meters(entry["coords"], coords) <= radius_meters:
            return entry["mph"]
    return None

File: js/math/speedLimitParsing.js
// Pure OSM maxspeed-tag parsing -- no network, no DOM. Mirrored in
// python/paceometer_math/speed_limit_parsing.py (see tests/golden_vectors/).
import { haversineMeters } from "./geoMath.js";

// OSM maxspeed values show up as "25 mph", a bare "50" (US ways are tagged
// in mph explicitly; a bare number in this country means mph), "80 km/h", or
// non-numeric values like "national"/"none" that aren't a usable limit.
export function parseMaxspeedTag(raw) {
  if (typeof raw !== "string") return null;
  const trimmed = raw.trim().toLowerCase();
  const kmhMatch = trimmed.match(/^(\d+(?:\.\d+)?)\s*km\/?h$/);
  if (kmhMatch) return parseFloat(kmhMatch[1]) * 0.621371;
  const mphMatch = trimmed.match(/^(\d+(?:\.\d+)?)\s*mph$/);
  if (mphMatch) return parseFloat(mphMatch[1]);
  const bareMatch = trimmed.match(/^(\d+(?:\.\d+)?)$/);
  if (bareMatch) return parseFloat(bareMatch[1]);
  return null;
}

export function extractMaxspeedMph(data) {
  for (const element of (data && data.elements) || []) {
    const mph = parseMaxspeedTag(element.tags && element.tags.maxspeed);
    if (mph !== null) return mph;
  }
  return null;
}

// Most recent still-fresh, still-nearby confirmed reading in `cache`
// (an array of { coords, mph, timestamp }), or null if nothing qualifies.
// Walked newest-first so a stretch that's been requeried since first being
// cached returns its latest value.
export function cachedSpeedLimitNear(cache, coords, timestamp, maxAgeMs, radiusMeters) {
  for (let i = cache.length - 1; i >= 0; i--) {
    const entry = cache[i];
    if (timestamp - entry.timestamp > maxAgeMs) continue;
    if (haversineMeters(entry.coords, coords) <= radiusMeters) return entry.mph;
  }
  return null;
}
